src/compiler/emitter.ts

```diff
@@ -7086,7 +7112,21 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
             }
 
             function emitDetachedComments(node: TextRange) {
-                let leadingComments = getLeadingCommentRanges(currentSourceFile.text, node.pos);
+                let leadingComments: CommentRange[];
+                if (compilerOptions.removeComments) {
+                    // removeComments is true, only reserve pinned comment at the top of file
+                    // For example:
+                    //      /*! Pinned Comment */
+                    //
+                    //      var x = 10;
+                    leadingComments = node.pos === 0 ?
+                        filter(getLeadingCommentRanges(currentSourceFile.text, node.pos), isPinnedComments) : [];

```

Comment:
leading comments can be undefined here. we check for it later. no need for the empty array literal allocation